keps/0007-command-asserts.md

```diff
@@ -0,0 +1,103 @@
+owners:
+  - "@nfnt"
+creation-date: 2020-10-28
+last-updated: 2020-10-28
+status: provisional
+---
+
+# Assertions for CLI commands
+
+## Table of Contents
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+    - [User Story 1](#user-story-1)
+    - [User Story 2](#user-story-2)
+  - [Implementation Details](#implementation-details)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+
+## Summary
+
+By asserting CLI commands, kuttl can support many additional test scenarios that examing the internal state of an operator.
+
+## Motivation
+
+Operators and custom resources can be used to deploy complex applications. For example, KUDO provides operators for Apache Cassandra and Apache Kafka. Deploying custom resources of these operators may not always result in changed Kubernetes resources but also in changes that aren't surfaced by Kubernetes resources but are internal to the operator. Asserting these changes only works through accessing service APIs or running commands and parsing their output. kuttl should provide tools to assert such output.
+
+### Goals
+
+- Assert that commands examining the internal state of an operator succeed or fail
+- Keep eventual consistency when asserting commands
+
+### Non-Goals
+
+- Introduce "flows" to determine when to run commands. I.e. to run commands once before or after asserting resources
+
+## Proposal
+
+Introduce commands as part of `TestAssert`. These commands are run at a regular interval like resource assertions until they either succeed or a timeout of the assert is reached. If any of the commands fail, the assert is considered to have failed. A command's return value determines if a command failed or not.
+
+As a start, complex commands that parse output can be described using shell environments. At a later stage we can implement additional tools to simplify common tasks. E.g. to search with a regular expression in command output.
+
+### User Stories
+
+#### User Story 1
+
+The JVM options of a KUDO operator can be tuned using parameters. On deployment, the JVM options are logged in the pods running the application provided by the operator. To assert that a parameter change resulted in changed JVM options we need to run `kubectl logs pod-name` and search the logs for a specific string once the operator has been deployed.
+
+#### User Story 2
+
+An operator can be configured to encrypt it's API. To assert that the API endpoints are indeed encrypted we can run a specific `curl` command and check that it returns successfully once the operator has been deployed.
+
+### Implementation Details
+
+Command assertions use most fields of the existing commands for `TestStep` and follow their behavior. The `ignoreFailure`, `background` and `timeout` fields are removed as they are not needed here.
```

Comment:
I think we need at least the `timeout` field as well - We should certainly monitor the executed command to make sure it doesn't run forever, and if we do that we should have the `timeout`.

For the `background` - I'm not sure. Could be useful, especially for things like the `kubectl logs pod-name`, I'd might want to add the `-f` to follow the logs and don't fetch them repeatedly if I know that a certain line will show up at some point. For this a background task might be good, as it wouldn't stop all other commands.

Reply:
A `TestAssert` already has a timeout parameter. Individual timeouts on commands create a challenge with multiple commands. If, say, command A has a timeout of 10 seconds, command B has a timeout of 30 seconds and command B, we check the command status every second... should we stop checking the status of command A after it has succeeded in the first 10 seconds but we're still waiting for command B to succeed? Having a single timeout as part of `TestAssert` makes the implementation much simpler as we don't have to support these edge cases.

Reply:
@nfnt we likely need some clean up on timeouts perhaps it is just documentation.  There are inconsistencies with timeouts and how they are applied through kuttl.    if there is a TestAssert timeout (for ex. 30s).  Is it your expectation that 30s is the timeout for all of the Assert process... or for each command?

Reply:
Same behavior as before for `TestAssert`, this will be the loop:

```go
	for i := 0; i < timeout; i++ {
		// start fresh
		testErrors = []error{}
		for _, expected := range objects {
			testErrors = append(testErrors, s.CheckResource(expected, namespace)...)
		}

		for _, command := range commands {
			testErrors = append(testErrors, s.CheckCommand(command)...)
		}

		if len(testErrors) == 0 {
			break
		}

		time.Sleep(time.Second)
	}
```

As before, the timeout isn't a value in seconds, but number of iterations of this loop.